### PR TITLE
bpo-37931: crash on OSX re-initializing os.environ

### PR DESCRIPTION
**Pull request by benoithudson**
On most platforms, the `environ` symbol is accessible everywhere.

In a dylib on OSX, it's not easily accessible, you need to find it with
_NSGetEnviron.

The code was caching the *value* of environ. But a setenv can change the value,
leaving garbage at the old value. Fix: don't cache the value of environ, just
read it every time.

https://bugs.python.org/issue37931

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by ned-deily** [comment]
@ronaldoussoren should review this

File: Modules/posixmodule.c
Comment (ronaldoussoren) [inline Modules/posixmodule.c:1376]: It is possible to simplify the rest of the code by doing:

`#define environ (*_NSGetEnviron())`

Then the new block at line 1398 can be dropped entirely.

File: Modules/posixmodule.c
Comment (ronaldoussoren) [inline Modules/posixmodule.c:1398]: See my previous comment. This block can be removed when using the \#define I mention there.

**Review by ronaldoussoren — COMMENTED** [review]
The patch looks good, but its possible to centralise the macOS specific code a bit more, see my comments below.

File: Misc/NEWS.d/next/macOS/2019-08-23-12-14-34.bpo-37931.goYgQj.rst
Comment (vstinner) [inline Misc/NEWS.d/next/macOS/2019-08-23-12-14-34.bpo-37931.goYgQj.rst]: You should mention Benoît Hudson.

File: Modules/posixmodule.c
Comment (vstinner) [reply Modules/posixmodule.c:1398]: I prefer to see an explicit _NSGetEnviron() rather than macros black magic.

Comment (ronaldoussoren) [reply Modules/posixmodule.c:1398]: Neither option is obviously better. Just calling _NSGetEnviron does make it more explicit what's going on.

File: Modules/posixmodule.c
Comment (ronaldoussoren) [reply Modules/posixmodule.c:1376]: See resolved comments below. Victor prefers an explicit call to _NSGetEnvironment.

Comment (benoithudson) [reply Modules/posixmodule.c:1376]: Given the other `#ifdef` handling around there I agree with Victor.

**Comment by benoithudson** [comment]
How does backporting work -- do I prepare a new PR for those or does someone else take care of it?

I'd ideally like to backport to 2.7 since that's where I actually hit the bug. This part of the codebase hasn't changed in a *long* time.

**Comment by ned-deily** [comment]
@benoithudson, backporting is attempted automatically when a core developer adds the appropriate "needs backport to ..." labels to the PR, as @ronaldoussoren has done. It could be considered for 2.7.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by benoithudson** [comment]
I can't add labels; if someone can add `needs backport to 2.7` that would be great.

**Review by ronaldoussoren — APPROVED** [review]
LGTM

**Comment by ronaldoussoren** [comment]
That's annoying: "squash and merge" fails for me at the moment, and twice in a row ("Merge attempt failed"). Will try again later.

**Comment by benoithudson** [comment]
Any luck trying again?

**Comment by vstinner** [comment]
The change looks perfectly safe to me. I cannot test it, but I trust @benoithudson managed to confirm that his change fix his crash.

**Comment by vstinner** [comment]
@benoithudson: The automated backport to 2.7 and 3.7 failed. Would you mind to attempt to backport manually the change starting with "git cherry-pick -x 723f71abf7ab0a7be394f9f7b2daa9ecdf6fb1eb" or use cherry_picker, as explained in previous comments.

**Comment by benoithudson** [comment]
Sure, I'll try to do that this weekend.

**Pull request merged**
